encryption.go

```diff
@@ -0,0 +1,413 @@
+		}
+	}
+	return r
+}
+
+func (e *Encryption) decryptResponse(response *http.Response, host string) ([]byte, error) {
+	body, _ := ioutil.ReadAll(response.Body)
+	parts := deleteEmpty(bytes.Split(body, []byte(MimeBoundary+"\r\n")))
+	var message []byte
+
+	for i := 0; i < len(parts); i += 2 {
+		header := parts[i]
+		payload := parts[i+1]
+
+		expectedLengthStr := bytes.SplitAfter(header, []byte("Length="))[1]
+		expectedLength, _ := strconv.Atoi(string(bytes.TrimSpace(expectedLengthStr)))
+
+		// remove the end MIME block if it exists
+		if bytes.HasSuffix(payload, []byte(MimeBoundary+"--\r\n")) {
+			payload = payload[:len(payload)-BoundaryLength-4]
+		}
+		encryptedData := bytes.ReplaceAll(payload, []byte("\tContent-Type: application/octet-stream\r\n"), []byte{})
+		decryptedMessage, err := e.decryptMessage(encryptedData, host)
+		if err != nil {
+			return nil, err
+		}
+
+		actualLength := int(len(decryptedMessage))
+		if actualLength != expectedLength {
+			return nil, errors.New("encrypted length from server does not match the expected size, message has been tampered with")
+		}
+
+		message = append(message, decryptedMessage...)
+	}
+
+	return message, nil
+}
+
+func (e *Encryption) decryptMessage(encryptedData []byte, host string) ([]byte, error) {
+	switch e.protocol {
+	case "ntlm":
+		return e.decryptNtlmMessage(encryptedData, host)
+	case "credssp":
+		return e.decryptCredsspMessage(encryptedData, host)
+	case "kerberos":
+		return e.decryptKerberosMessage(encryptedData, host)
+	default:
+		return nil, errors.New("Encryption for protocol " + e.protocol + " not supported in pywinrm")
+	}
+}
+
+func (e *Encryption) decryptNtlmMessage(encryptedData []byte, host string) ([]byte, error) {
+	signatureLength := int(binary.LittleEndian.Uint32(encryptedData[:4]))
+	signature := encryptedData[4 : signatureLength+4]
+	encryptedMessage := encryptedData[signatureLength+4:]
+
+	message, err := e.ntlmClient.SecuritySession().Unwrap(encryptedMessage, signature)
+	if err != nil {
+		return nil, err
+	}
+	return message, nil
+}
+
+func (e *Encryption) decryptCredsspMessage(encryptedData []byte, host string) ([]byte, error) {
+	/*
+		TODO
+		encryptedMessage := encryptedData[4:]
+
+		credsspContext, ok := e.session.Auth.Contexts()[host]
+		if !ok {
+			return nil, fmt.Errorf("credssp context not found for host: %s", host)
+		}
+
+		message, err := credsspContext.Unwrap(encryptedMessage)
+		if err != nil {
+			return nil, err
+		}
+		return message, nil
+	*/
+	return nil, nil
+}
+
+func (enc *Encryption) decryptKerberosMessage(encryptedData []byte, host string) ([]byte, error) {
+	/*
+		TODO
+		signatureLength := binary.LittleEndian.Uint32(encryptedData[0:4])
+		signature := encryptedData[4 : 4+signatureLength]
+		encryptedMessage := encryptedData[4+signatureLength:]
+
+		message, err := enc.session.Auth.UnwrapWinrm(host, encryptedMessage, signature)
+		if err != nil {
+			return nil, err
+		}
+
+		return message, nil
+	*/
+	return nil, nil
+}

```

Comment:
I'd prefer that we error instead of those unimplemented functions

Reply:
commented out for future work